crates/bvs-vault-router/src/contract.rs

```diff
@@ -325,13 +326,149 @@ mod execute {
+        if operator_service_status != u8::from(RegistrationStatus::Active) {
+            return Err(InvalidSlashingRequest {
+                msg: "Service and Operator are not active at timestamp.".to_string(),
+            });
+        }
+
+        let now = env.block.time;
+        let cond_expired = now > slash_req.request_expiry;
+        let cond_not_aged = now
+            < slash_req
+                .request_time
+                .plus_seconds(slashing_parameters.resolution_window);
+
+        if cond_expired {
+            return Err(ContractError::InvalidSlashingRequest {
+                msg: "Slash id is expired".to_string(),
+            });
+        };
+
+        if cond_not_aged {
+            return Err(ContractError::InvalidSlashingRequest {
+                msg: "Resolution window for this id has not passed".to_string(),
+            });
+        };
+
+        let vaults_managed = state::OPERATOR_VAULTS
+            .prefix(&accused_operator)
+            .keys(deps.storage, None, None, cosmwasm_std::Order::Ascending)
+            .collect::<StdResult<Vec<_>>>()?;
+
+        // Array instead of map to avoid the overheads
+        let mut messages = Vec::new();
+        for vault in vaults_managed {
+            let vault_info = vault::get_vault_info(deps.as_ref(), &vault)?;
+            let total_assets = vault_info.total_assets;
+            let bips = Uint128::from(slash_req.request.bips as u128);
+
+            let slash_percent = bips.checked_div(Uint128::from(100_u128)).unwrap();
+
+            // multiply_ratio is always floored.
+            let slash_absolute =
+                total_assets.multiply_ratio(slash_percent, Uint128::from(100_u128));
+
+            SLASH_LOCKED.save(deps.storage, (id.clone(), &vault), &slash_absolute)?;
+
+            // Can't slash lock a zero asset vault the vault will reject.
+            // But we still need to populate the `SLASH_LOCKED` map to indicate
+            // slash locking for given id is already done.
+            if slash_absolute.is_zero() {
+                continue;
+            };
+
+            let amount = vault::Amount(slash_absolute);
+            let msg = vault::ExecuteMsg::SlashLocked(amount);
+            let exec_msg = cosmwasm_std::WasmMsg::Execute {
+                contract_addr: vault.to_string(),
+                msg: to_json_binary(&msg)?,
+                funds: vec![],
+            };
+
+            messages.push(exec_msg);
+        }
+
+        let response = Response::new().add_event(
+            Event::new("SlashLocked")
+                .add_attribute("service", info.sender)
+                .add_attribute("operator", accused_operator)
+                .add_attribute("slashing_request_id", id.to_string())
+                .add_attribute("bips", slash_req.request.bips.to_string())
+                .add_attribute("affected_vaults", messages.len().to_string()),
+        );
+
+        Ok(response.add_messages(messages))
```

Comment:
```suggestion
        Ok(Response::new()
            .add_messages(messages)
            .add_event(
            Event::new("SlashLocked")
                .add_attribute("service", info.sender)
                .add_attribute("operator", accused_operator)
                .add_attribute("slashing_request_id", id.to_string())
                .add_attribute("bips", slash_req.request.bips.to_string())
                .add_attribute("affected_vaults", messages.len().to_string()),
        ))
```

I find the `let response` declaration redundant here.